src/state.py

```diff
@@ -735,138 +747,139 @@ def recompute_pass_from_possession(self):
-                    distance_list.append((player_id, dist))
+    def recompute_possessions_v2(self):
+        """
+        Calculate and store the frame-by-frame score for each player based on their possession position.
+        This score is stored in a rolling window covering the last 50 frames.
+        """
+        # Initialize an empty list for rolling scores
+        self.rolling_scores = []
 
-            # Sort by adjusted distance, closest first
-            distance_list.sort(key=lambda x: x[1])
-            frame.possessions = [
-                player_id for player_id, _ in distance_list][:3]
+        for frame in self.frames:
+            # Dictionary to hold scores for each player in this frame
+            frame_scores = {}
 
-            # Update possession frequency and recent possessions
-            for player_id in frame.possessions:
-                possession_frequency[player_id] = possession_frequency.get(
-                    player_id, 0) + 1
-            recent_possessions.append(set(frame.possessions))
+            # Assign points based on possession position (1 for 1st, 0.75 for 2nd, 0.25 for 3rd)
+            for i, player_id in enumerate(frame.possessions[:2]):
+                points = [1, 0.75, 0.25][i]
+                frame_scores[player_id] = frame_scores.get(
+                    player_id, 0) + points
 
-            print(f"Frame {index}: {frame.possessions}")
+            # Append the scores of this frame to the rolling window
+            self.rolling_scores.append(frame_scores)
 
-    #TODO replace current recompute possesions with this more nuanced function
-    def compute_possession_intervals_AFK(self, window_size=30, min_interval_length=20, possession_threshold=0.8):
-        possession_history = deque(maxlen=window_size)
-        possession_intervals = []
-        current_possession = None
-
-        for frame_index, frame in enumerate(self.frames):
-            # Update possession history with weighted points
-            if frame.possessions:
-                # Assign points: 1 for first, 0.75 for second, 0.5 for third
-                for i, possessor in enumerate(frame.possessions[:3]):
-                    points = [1, 0.75, 0.5][i]
-                    possession_history.append((possessor, points))
+    def recompute_possessions_v3(self):
+        """
+        Determine the dominant player in possession for each frame based on the rolling window of the last 50 frames.
+        """
+        # Initialize an empty list to store the dominant player for each frame
+        self.dominant_possessions = []
+
+        for i in range(len(self.frames)):
+            # Dictionary to sum up scores in the rolling window for each player
+            rolling_sum = {}
+
+            # Accumulate scores for each player in the rolling window of the last 50 frames
+            for j in range(max(0, i - 49), i + 1):
+                for player_id, score in self.rolling_scores[j].items():
```

Comment:
This is pretty inefficient, usually when one computes a rolling windows, you continuously shift the window over by 1, adding the next frame's info while removing the last frame's info.  Can you make this more efficient. Another implementation you might consider is an exponential moving average